Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/CachedQueries/Driver/GoHomeRequest.hs

```diff
@@ -0,0 +1,143 @@
+{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
+
+module Storage.CachedQueries.Driver.GoHomeRequest where
+
+import Control.Monad
+import Data.Text (pack)
+import Domain.Types.Driver.GoHomeFeature.DriverGoHomeRequest as DDGR
+import Domain.Types.Driver.GoHomeFeature.DriverHomeLocation as DDHL
+import Domain.Types.GoHomeConfig
+import Domain.Types.Merchant
+import qualified Domain.Types.Person as DP
+import Kernel.Prelude
+import Kernel.Storage.Esqueleto.Config (EsqDBFlow)
+import qualified Kernel.Storage.Hedis as Hedis
+import Kernel.Types.App (MonadFlow)
+import Kernel.Types.CacheFlow
+import Kernel.Types.Common (MonadTime (getCurrentTime), generateGUID)
+import Kernel.Types.Id (Id)
+import Kernel.Types.SlidingWindowCounters (PeriodType (Days))
+import Kernel.Utils.Common (addUTCTime, fromMaybeM, getLocalCurrentTime)
+import Kernel.Utils.Logging (logDebug)
+import Kernel.Utils.SlidingWindowCounters (incrementPeriod)
+import qualified Storage.CachedQueries.GoHomeConfig as CQGHC
+import Storage.CachedQueries.Merchant.TransporterConfig as CQTC
+import Storage.Queries.Driver.GoHomeFeature.DriverGoHomeRequest as QDGR
+import Storage.Queries.Ride as Ride
+import Tools.Error (GenericError (..))
+
+makeGoHomeReqKey :: Id DP.Driver -> Text
+makeGoHomeReqKey = pack . ("CachedQueries:GoHomeRequest-driverId:" <>) . show
+
+getDriverGoHomeRequestInfo :: (CacheFlow m r, MonadFlow m, EsqDBFlow m r) => Id DP.Driver -> Id Merchant -> Maybe GoHomeConfig -> m CachedGoHomeRequest
+getDriverGoHomeRequestInfo driverId merchantId goHomeCfg = do
+  ghCfg <- maybe (CQGHC.findByMerchantId merchantId) return goHomeCfg
+  let initCnt = ghCfg.startCnt
+  expTime <- fromIntegral <$> asks (.cacheConfig.configsExpTime)
+  let ghkey = makeGoHomeReqKey driverId
+  currTime <- getLocalCurrentTime =<< ((CQTC.findByMerchantId merchantId >>= fromMaybeM (InternalError "Transporter config for timezone not found")) <&> (.timeDiffFromUtc))
+  Hedis.safeGet ghkey >>= \case
+    Just ghrData -> do
+      if maybe False (\validTill -> validTill > currTime || ghrData.isOnRide) ghrData.validTill
+        then return ghrData
+        else checkInvalidReqData ghrData currTime ghkey driverId merchantId ghCfg expTime
+    Nothing -> do
+      Hedis.setExp ghkey (templateGoHomeData Nothing initCnt Nothing Nothing False Nothing currTime) expTime
+      return $ templateGoHomeData Nothing initCnt Nothing Nothing False Nothing currTime
```

Comment:
we are initiating goHomeReqId as `Nothing`, and i can see at a lot of places you have validation on goHomeReqId not there then throw error. this looks risky can you avoid sending dummy values from this funciton, fields like Id and all shouldn't be Maybe.

Reply:
But if go home is not enabled what should I set it? Also, This design we both discussed and agreed upon? Also this is getting set using simple logic if enabled make a query to db and use the generated id to set it and if we have a go home request without a id that does not make any sense.

Reply:
> if we have a go home request without a id that does not make any sense

agreed, that why m saying don't even create this dummy object with nothing everywhere and some fields set to defaults. For now you have handled everywhere, but in future this may lead to wrong interpretation of what this function is doing and someone might use it wrongly.

Reply:
But whole point of creating this was to keep count in the redis and expire it every day. So its not dummy data count is there and reference time is keeping track which day's count it is. count is needed everywhere very frequently be it to dispaly or be it to show whether driver is eligible for go to request. So even when ghrid is null go to count will be there which is not dummy its an important parameter which is to be kept track of and expired/reset as per requirement time to time. Infact count was the reason you suggested to use redis you remember?

Reply:
sure, then lets change the type of object that we are storing to something like this:

```
data CachedGoHomeRequest = CachedGoHomeRequest
    {  goHomeRequest: Maybe GoHomeReqInfo
    ,  goHomeMeta: GoHomeMeta
    , status :: DriverGoHomeRequestStatus -> (Active | Inactive | Success | Failure)
    }
data GoHomeReqInfo = GoHomeReqInfo
    { driverGoHomeRequestId :: Id DriverGoHomeRequest
    , validTill :: UTCTime
    }
data GoHomeMeta = GoHomeMeta 
    {  count :: Int
    ,  goHomeReferenceTime :: UTCTime
    }
```

Reply:
Just sharing my perspective, here success and failure has no role this DS is for storing count, reqId and validity related info whether go to is success or failure same will be tracked in the DB. And here also if I want to access Go Home Request Id I have to do fromMaybeM only which was the issue as you were saying. In new one Like goHomeReqId and validTill here will be nothing (together) when status is not active and in the current one they will be nothing separately . I suggest lets just add proper commenting in the existing type so that in future if someone is reading it will have complete context. Also this data is being used very often throughout the code, I can restructure but that will be time consuming and Maybe affect is still there also now extra get fields will have to handle. Now it's up to you should we keep the current with proper commenting added in the type explaining meaning of each param or should we add this type, Not sure whether benefits will be worth the effort.

Reply:
Even if there is no significance in this data when its in redis the haskell type will have these constructors, otherwise how will you add them to DB and regarding the type change, i suggest to go with the types i proposed as it's more readable and adding inactive as a constructor will make it more explicit that go home is enabled or not. Also having few dependent fields as nothing separately can create issues in future, i agree it's handled rn, but if id and validTill can't exist without each other then ideally they both should he nothing or both should have a value that is ensured by the type i suggested this way you will not have to handle if validity exist and id doesn't etc kind of weird cases that should never come and if they came it would be hard to debug with the current type. If its very time consuming then maybe we can create another issue for refactoring and till then you can resolve other comments. @hkmangla  your views on this....should we do now or later ? Later we may have to backfill data from older type in redis to newer type if released during daytime or add backwards compatibility code...